File: Solution-Java/Design_622.java
class MyCircularQueue {
    
    private int[] q;
    private int cap, size;
    private int start, end;

    public MyCircularQueue(int k) {
        q = new int[k];
        cap = k;
        size = 0;
        start = 0;
        end = 0;
    }
    
    public boolean enQueue(int value) {
        if (isFull()) return false;
        q[end] = value;
        end = (end + 1) % cap;
        size++;
        return true;
    }
    
    public boolean deQueue() {
        if (isEmpty()) return false;
        start = (start + 1) % cap;
        size--;
        return true;
    }
    
    public int Front() {
        if (isEmpty()) return -1;
        return q[start];
    }
    
    public int Rear() {
        if (isEmpty()) return -1;
        return q[(end + cap - 1) % cap];
    }
    
    public boolean isEmpty() {
        return size == 0;
    }
    
    public boolean isFull() {
        return size == cap;
    }
}

File: Solution-Python/Design_622.py
class MyCircularQueue:

    def __init__(self, k: int):
        self.q = [0] * k
        self.cap = k
        self.size = 0
        self.start = 0
        self.end = 0
        
    def enQueue(self, value: int) -> bool:
        if self.isFull():
            return False
        self.q[self.end] = value
        self.end = (self.end + 1) % self.cap
        self.size += 1
        return True

    def deQueue(self) -> bool:
        if self.isEmpty():
            return False
        self.start = (self.start + 1) % self.cap
        self.size -= 1
        return True

    def Front(self) -> int:
        if self.isEmpty():
            return -1
        return self.q[self.start]

    def Rear(self) -> int:
        if self.isEmpty():
            return -1
        idx = (self.end + self.cap - 1) % self.cap
        return self.q[idx]

    def isEmpty(self) -> bool:
        return self.size == 0

    def isFull(self) -> bool:
        return self.size == self.cap